### PR TITLE
[13.0][OU-IMP] l10n_be: migration script (account.account.tag)

### PR DESCRIPTION
**Pull request by remytms**
Migrate account.account.tag to fit the new account.tax and its repartition lines.

### CODE REVIEW

**Comment by remytms** [comment]
@sbidoul  this PR might interest you when migrating Belgian accounting. :)

**Comment by remytms** [comment]
@marielejeune Can you give a look at this migration script? Thanks. :)

**Comment by marielejeune** [comment]
Hi @remytms, I've seen your PR a few days ago, but I'm really not an accounting specialist and I don't think I've the competencies to functionally give any comment :/

**Comment by remytms** [comment]
Thanks @marielejeune , if you know a Belgian accounting person do not hesitate to ping him/her.

**Comment by remytms** [comment]
@luc-demeyer You may be interested by this pull request. :)

**Comment by marielejeune** [comment]
Hi @remytms 
I am trying your script on a customer's database migration. I'll give you a complete feedback when I'll finish the migration. But I already have a functional question, since I'm not an accounting specialist. I got the following error: 

There is several account.tax found for the account.tax.template ID 141 '21% EU S.' that are used on account.move.line. Clean up account.tax by renaming your specific account.tax and migrate it via a dedicated script. The matching account.tax ID 133, 551

Indeed, both taxes with ID 133 and 551 have the same name, look the same, but one of them is linked to Company 1 (Belgian company) and the second one is linked to Company 2 (also Belgian company).

Is it a problem to have 2 taxes coming from the same tax template? If yes, what should have been done when initializing invoicing for these 2 companies? Duplicate the tax templates?

**Comment by fd-oncodna** [comment]
Indeed as @marielejeune said I think this does not work if you have multiple companies in Belgium, as the taxes will be duplicated - but not the tax templates, as they are created through XML. You should loop on the companies, change the signature of `find_account_tax(env, xmlid, account_tax_template)` to `find_account_tax(env, xmlid, account_tax_template, company)` and search both the taxes and the move lines restricted to `company`.

File: addons/l10n_be/migrations/13.0.2.0/post-migration.py [hunk 1]
```diff
@@ -0,0 +1,700 @@
...
+        SELECT account_tax_id
+        FROM account_move_line_account_tax_rel
+        WHERE account_tax_id = %s
+        """,
+        (account_tax.id,),
+    )
+    return len(env.cr.fetchall()) > 0
+
+
+def find_account_tax(env, xmlid, account_tax_template):
```
Comment (marielejeune) [inline addons/l10n_be/migrations/13.0.2.0/post-migration.py]:
```suggestion
def find_account_tax(env, xmlid, account_tax_template, company_id):
```

File: addons/l10n_be/migrations/13.0.2.0/post-migration.py [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+    """Return the account tax based on the given account.tax.template"""
+    current_name = account_tax_template.name
+    old_names = get_old_account_tax_names(xmlid)
+    domain = [("type_tax_use", "=", account_tax_template.type_tax_use)]
```
Comment (marielejeune) [inline addons/l10n_be/migrations/13.0.2.0/post-migration.py]:
```suggestion
    domain = [("type_tax_use", "=", account_tax_template.type_tax_use), ("company_id", "=", company_id)]
```

File: addons/l10n_be/migrations/13.0.2.0/post-migration.py [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+
+        if not account_tax:
+            continue
+
+        apply_new_tax_tags_on_account_tax(
+            env, account_tax, account_tax_template
+        )
+
+        # Rename account_tax
+        account_tax.name = account_tax_template.name
```
Comment (marielejeune) [inline addons/l10n_be/migrations/13.0.2.0/post-migration.py]: Loop on companies
```suggestion
    for company in env["res.company"].search([]):
        for xmlid, account_tax_template in get_account_tax_templates_iterator(env):
            account_tax = find_account_tax(env, xmlid, account_tax_template, company.id)

            if not account_tax:
                continue

            apply_new_tax_tags_on_account_tax(
                env, account_tax, account_tax_template
            )

            # Rename account_tax
            account_tax.name = account_tax_template.name
```

File: addons/l10n_be/migrations/13.0.2.0/post-migration.py [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+            lambda r: r.move_id.type in ("out_refund", "in_refund")
+        )
+
+    # Adding account.move.line that are linked to this specific
+    # repartition_line
+    account_move_line_ids += env["account.move.line"].search(
+        [("tax_repartition_line_id", "=", repartition_line.id)]
+    )
+
+    if not account_move_line_ids:
```
Comment (marielejeune) [inline addons/l10n_be/migrations/13.0.2.0/post-migration.py]:
```suggestion
    if not account_move_line_ids or not old_tag_ids:
```

**Review by marielejeune — CHANGES_REQUESTED** [review]
I've tried to run your script on a DB with 2 Belgian companies, hence I've adapted it to loop on companies as @fd-oncodna suggested. Here are my comments, please have a look and tell me if it seems good to you:

**Comment by remytms** [comment]
@marielejeune @fd-oncodna Thanks a lot for your remark and suggestion about multi-company. Changes added.

**Comment by remytms** [comment]
I close this PR in favor of this one: #4432

**Pull request closed without merging**